### PR TITLE
Add bidirectional streaming for evidences and assessment results

### PR DESCRIPTION
**Pull request by garuppel**

### CODE REVIEW

File: service/orchestrator/orchestrator.go [hunk 1]
```diff
@@ -83,6 +81,7 @@ func init() {
 // ServiceOption is a function-style option to configure the Orchestrator Service
 type ServiceOption func(*Service)
 
+// TODO(all): Function currently not used
 // WithMetricsFile can be used to load a different metrics file
 func WithMetricsFile(file string) ServiceOption {
 	return func(s *Service) {
```
Comment (garuppel) [inline service/orchestrator/orchestrator.go:87]:
```suggestion

```

File: service/orchestrator/orchestrator.go [hunk 2]
```diff
@@ -174,10 +173,6 @@ func (s *Service) GetMetricConfiguration(_ context.Context, req *orchestrator.Ge
 // configuration for a particular metric within the service, the default metric configuration is
 // inserted into the list.
 func (s *Service) ListMetricConfigurations(ctx context.Context, req *orchestrator.ListMetricConfigurationRequest) (response *orchestrator.ListMetricConfigurationResponse, err error) {
-	response = &orchestrator.ListMetricConfigurationResponse{
-		Configurations: make(map[string]*assessment.MetricConfiguration),
-	}
-
```
Comment (garuppel) [inline service/orchestrator/orchestrator.go:180]: Add:	response = &orchestrator.ListMetricConfigurationResponse{
		Configurations: make(map[string]*assessment.MetricConfiguration),
	}

File: service/orchestrator/orchestrator.go [hunk 3]
```diff
@@ -205,40 +200,57 @@ func (s *Service) StoreAssessmentResult(_ context.Context, req *orchestrator.Sto
 
 		go s.informHook(nil, newError)
 
-		return resp, status.Errorf(codes.InvalidArgument, "invalid req: %v", err)
+		resp = &orchestrator.StoreAssessmentResultResponse{
+			Status:        false,
+			StatusMessage: newError.Error(),
+		}
+
+		return resp, status.Errorf(codes.InvalidArgument, "%v", err)
```
Comment (garuppel) [inline service/orchestrator/orchestrator.go:208]:
```suggestion
		return resp, status.Errorf(codes.InvalidArgument, "invalid assessment result")
```

File: service/orchestrator/orchestrator.go [hunk 4]
```diff
@@ -205,40 +200,57 @@ func (s *Service) StoreAssessmentResult(_ context.Context, req *orchestrator.Sto
...
+		if err != nil {
+			log.Errorf("Error storing assessment result: %v", err)
 		}
 
-		_, err = s.StoreAssessmentResult(context.Background(), storeAssessmentResultReq)
+		err = stream.Send(res)
 		if err != nil {
-			return err
+			log.Errorf("Error when response was sent to the client: %v", res)
+			return status.Errorf(codes.Unknown, "cannot send stream response: %v", err)
```
Comment (garuppel) [inline service/orchestrator/orchestrator.go:253]:
```suggestion
			return status.Errorf(codes.Unknown, "cannot stream response to the client: %v", err)
```

File: service/orchestrator/orchestrator_test.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021-2022 Fraunhofer AISEC
+// Copyright 2021 Fraunhofer AISEC
```
Comment (garuppel) [inline service/orchestrator/orchestrator_test.go:1]:
```suggestion
// Copyright 2021-2022 Fraunhofer AISEC
```

File: service/orchestrator/orchestrator_test.go [hunk 6]
```diff
@@ -41,14 +45,15 @@ import (
 	"google.golang.org/protobuf/types/known/timestamppb"
 
 	"clouditor.io/clouditor/api/assessment"
+
```
Comment (garuppel) [inline service/orchestrator/orchestrator_test.go:48]:
```suggestion
```

**Review by garuppel — COMMENTED** [review]
Fix more merge errors

File: service/orchestrator/orchestrator.go (hunk 1 above)
Comment (oxisto) [reply service/orchestrator/orchestrator.go:87]: I would keep this function please. I need it in the future :)

Comment (garuppel) [reply service/orchestrator/orchestrator.go:87]: @oxisto Do we need that?

Comment (garuppel) [reply service/orchestrator/orchestrator.go:87]: Ok

File: service/assessment/assessment_test.go [hunk 7]
```diff
@@ -238,53 +245,66 @@ func TestAssessEvidences(t *testing.T) {
 		UnimplementedAssessmentServer assessment.UnimplementedAssessmentServer
 	}
 	type args struct {
-		stream assessment.Assessment_AssessEvidencesServer
+		streamToServer *mockAssessmentServerStream
```
Comment (oxisto) [inline service/assessment/assessment_test.go:248]: In the arguments it would be better to use the interface `assessment.Assessment_AssessEvidencesServer` rather then the implementation

File: service/orchestrator/orchestrator_test.go [hunk 8]
```diff
@@ -41,14 +45,15 @@ import (
...
 	"clouditor.io/clouditor/api/orchestrator"
 	"clouditor.io/clouditor/persistence/inmemory"
 	"github.com/stretchr/testify/assert"
 )
 
 const (
 	assessmentResultID1 = "11111111-1111-1111-1111-111111111111"
-	assessmentResultID2 = "11111111-1111-1111-1111-111111111112"
+	//assessmentResultID2 = "11111111-1111-1111-1111-111111111112"
```
Comment (oxisto) [inline service/orchestrator/orchestrator_test.go:56]: ?

File: service/orchestrator/orchestrator_test.go [hunk 9]
```diff
@@ -249,19 +261,59 @@ func TestStoreAssessmentResult(t *testing.T) {
...
+		count2 = 2
+	)
+
+	type fields struct {
+		countElementsInMock    int
+		countElementsInResults int
+	}
+
 	type args struct {
-		stream orchestrator.Orchestrator_StoreAssessmentResultsServer
```
Comment (oxisto) [inline service/orchestrator/orchestrator_test.go:253]: use interface please

**Review by oxisto — COMMENTED** [review]
Minor things in the tests

File: service/orchestrator/orchestrator_test.go (hunk 8 above)
Comment (garuppel) [reply service/orchestrator/orchestrator_test.go:56]: I forgot to delete that.

File: service/assessment/assessment.go [hunk 10]
```diff
@@ -166,56 +165,69 @@
...
+		}
 		if err != nil {
-			// If no more input of the stream is available, return SendAndClose `error`
-			if err == io.EOF {
-				log.Infof("Stopped receiving streamed evidences")
-				return stream.SendAndClose(&emptypb.Empty{})
-			}
-
-			return err
+			log.Errorf("Assessment: Cannot receive stream request: %v", err)
```
Comment (lebogg) [inline service/assessment/assessment.go:213]:
```suggestion
			log.Errorf("Cannot receive stream request: %v", err)
```

Comment (lebogg) [reply service/assessment/assessment.go:213]: If omitted, also consider the following structure to avoid duplicated strings:
```
			err = fmt.Errorf("cannot receive stream request: %v", err)
			log.Error(err)
			return status.Error(codes.Unknown, err.Error())
```

File: service/assessment/assessment.go [hunk 11]
```diff
@@ -228,7 +240,7 @@
 	evaluations, err := policies.RunEvidence(evidence, s)
 	if err != nil {
 		newError := fmt.Errorf("could not evaluate evidence: %w", err)
-		log.Errorf(newError.Error())
+		log.Error(newError.Error())
```
Comment (lebogg) [inline service/assessment/assessment.go:243]:
```suggestion
		log.Error(newError)
```

File: service/evidence/evidence_store.go [hunk 12]
```diff
@@ -95,28 +95,36 @@
...
+		}
 		if err != nil {
-			// If no more input of the stream is available, return SendAndClose `error`
-			if err == io.EOF {
-				log.Infof("Stopped receiving streamed evidences")
-				return stream.SendAndClose(&emptypb.Empty{})
-			}
-
-			return err
+			log.Errorf("EvidenceStore: Cannot receive stream request: %v", err)
```
Comment (lebogg) [inline service/evidence/evidence_store.go:111]: Same as for assessment: Omit EvidenceStore and consider to avoid duplicate string

File: service/orchestrator/metrics.go [hunk 13]
```diff
@@ -60,12 +60,13 @@ func (*Service) CreateMetric(_ context.Context, req *orchestrator.CreateMetricRe
 	// Validate the metric request
 	err = req.Metric.Validate(assessment.WithMetricRequiresId())
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "Validation of metric failed: %v", err)
+		log.Errorf("Validation of metric failed: %v", err)
+		return nil, status.Error(codes.InvalidArgument, "validation of metric failed")
```
Comment (lebogg) [inline service/orchestrator/metrics.go:64]: Why removing the error message which contains the specific part of the validation error?

Comment (lebogg) [reply service/orchestrator/metrics.go:64]: Here we could also avoid writing same string twice

File: service/orchestrator/metrics.go [hunk 14]
```diff
@@ -60,12 +60,13 @@ func (*Service) CreateMetric(_ context.Context, req *orchestrator.CreateMetricRe
...
 	}
 
 	// Check, if metric id already exists
 	if _, ok := metricIndex[req.Metric.Id]; ok {
-		return nil, status.Error(codes.AlreadyExists, "Metric with identifer already exists")
+		return nil, status.Error(codes.AlreadyExists, "metric with identifier already exists")
```
Comment (lebogg) [inline service/orchestrator/metrics.go:69]:
```suggestion
		return nil, status.Error(codes.AlreadyExists, "metric already exists")
```

File: service/orchestrator/metrics.go [hunk 15]
```diff
@@ -85,12 +86,14 @@ func (*Service) UpdateMetric(_ context.Context, req *orchestrator.UpdateMetricRe
 	// Validate the metric request
 	err = req.Metric.Validate()
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "Validation of metric failed: %v", err)
+		log.Errorf("Validation of metric failed: %v", err)
+		return nil, status.Errorf(codes.InvalidArgument, "validation of metric failed: %v", err)
```
Comment (lebogg) [inline service/orchestrator/metrics.go:90]: duplicated string

File: service/orchestrator/metrics.go [hunk 16]
```diff
@@ -85,12 +86,14 @@ func (*Service) UpdateMetric(_ context.Context, req *orchestrator.UpdateMetricRe
...
 	}
 
 	// Check, if metric exists according to req.MetricId
 	if metric, ok = metricIndex[req.MetricId]; !ok {
-		return nil, status.Error(codes.NotFound, "Metric with identifer does not exist")
+		log.Errorf("Metric with identifier %s does not exist", req.MetricId)
+		return nil, status.Errorf(codes.NotFound, "metric with identifier %s does not exist", req.MetricId)
```
Comment (lebogg) [inline service/orchestrator/metrics.go:96]: duplicated string

File: service/orchestrator/orchestrator.go [hunk 17]
```diff
@@ -195,8 +193,6 @@
 // StoreAssessmentResult is a method implementation of the orchestrator interface: It receives an assessment result and stores it
 func (s *Service) StoreAssessmentResult(_ context.Context, req *orchestrator.StoreAssessmentResultRequest) (resp *orchestrator.StoreAssessmentResultResponse, err error) {
```
Comment (lebogg) [inline service/orchestrator/orchestrator.go:195]:
```suggestion
```

File: service/orchestrator/orchestrator.go [hunk 18]
```diff
@@ -205,40 +201,57 @@
...
-				return stream.SendAndClose(&emptypb.Empty{})
-			}
+		// If no more input of the stream is available, return
+		if err == io.EOF {
+			return nil
+		}
 
-			return err
+		if err != nil {
+			log.Errorf("Orchestrator: Cannot receive stream request: %v", err)
```
Comment (lebogg) [inline service/orchestrator/orchestrator.go:238]: See assessment and evidence store comment

**Review by lebogg — COMMENTED** [review]
~~For the unary RPCs, I would prefer to have an error response only. Maybe I don't see the benefit of having an additional response (which, for me, is stating the same as the error)?~~
We need response for status of message (e.g. `ASSESSED` or `WAITING_FOR_RELATED`).

File: service/assessment/assessment.go [hunk 19]
```diff
@@ -166,56 +165,69 @@
...
+// AssessEvidences is a method implementation of the assessment interface: It assesses multiple evidences (stream) and responds with a stream.
 func (s *Service) AssessEvidences(stream assessment.Assessment_AssessEvidencesServer) (err error) {
 	var (
 		req *assessment.AssessEvidenceRequest
+		res *assessment.AssessEvidenceResponse
 	)
 
 	for {
+
+		// TODO(all): Check context?
```
[the remark this answers is not in the document]
Comment (garuppel) [reply service/assessment/assessment.go:205]: This one: https://pkg.go.dev/context

File: service/assessment/assessment.go (hunk 10 above)
Comment (garuppel) [reply service/assessment/assessment.go:213]: The "problem" here is, that we usually capitalize the log messages and return errors in lower case. But we could also do it like that
`
err = fmt.Errorf("cannot receive stream request: %v", err)
log.Error(strings.Title(err.Error()))
return status.Error(codes.Unknown, err.Error())
`

File: service/assessment/assessment.go [hunk 20]
```diff
@@ -166,56 +165,69 @@
 		s.informHooks(nil, newError)
 
 		res = &assessment.AssessEvidenceResponse{
-			Status: false,
+			Status:        assessment.AssessEvidenceResponse_FAILED,
+			StatusMessage: newError.Error(),
 		}
 
-		return res, status.Errorf(codes.InvalidArgument, "invalid req: %v", newError)
+		return res, status.Error(codes.InvalidArgument, "invalid evidence")
```
[the remark this answers is not in the document]
Comment (garuppel) [reply service/assessment/assessment.go:172]: We could do it like that. The error message including the Validate()-error will be returned in the case of calling AssessEvidence() and in case of AssessEvidences() the error is stored in the statusMessage in the response. 

```suggestion
		return res, status.Error(codes.InvalidArgument, newError.Error())
```

File: service/orchestrator/orchestrator.go [hunk 21]
```diff
@@ -205,40 +201,57 @@ func (s *Service) StoreAssessmentResult(_ context.Context, req *orchestrator.Sto
...
```
Comment (garuppel) [inline service/orchestrator/orchestrator.go:237]:
```suggestion
			log.Errorf("Cannot receive stream request: %v", err)
```

File: api/assessment/assessment.proto [hunk 22]
```diff
@@ -77,7 +77,17 @@ message ConfigureAssessmentResponse {}
...
+message AssessEvidenceResponse {
+  enum AssessmentStatus {
+    ASSESSMENT_STATUS_UNSPECIFIED = 0;
+    WAITING_FOR_RELATED = 1;
+    ASSESSED = 2;
+    FAILED = 3;
+  }
+  AssessmentStatus status = 1;
+
+  string statusMessage = 2;
```
Comment (oxisto) [inline api/assessment/assessment.proto:89]: in protobuf style this should be `status_message`.

File: api/assessment/assessment.proto [hunk 23]
```diff
@@ -77,7 +77,17 @@ message ConfigureAssessmentResponse {}
 message TriggerAssessmentRequest { string someOption = 1; }
```
Comment (oxisto) [inline api/assessment/assessment.proto:77]: this should also be `some_option`

File: api/evidence/evidence_store.proto [hunk 24]
```diff
@@ -61,7 +58,10 @@ service EvidenceStore {
 }
 
 message StoreEvidenceRequest { Evidence evidence = 1; }
-message StoreEvidenceResponse { bool status = 1; }
+message StoreEvidenceResponse {
+  bool status = 1;
+  string statusMessage = 2;
```
Comment (oxisto) [inline api/evidence/evidence_store.proto:63]: `status_message`

File: logging/log.go [hunk 25]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2021 Fraunhofer AISEC
```
Comment (lebogg) [inline logging/log.go:1]:
```suggestion
// Copyright 2016-2022 Fraunhofer AISEC
```

File: api/orchestrator/orchestrator.proto [hunk 26]
```diff
@@ -239,7 +239,10 @@ message UpdateAssessmentToolRequest {
 message DeregisterAssessmentToolRequest { string tool_id = 1; }
 
 message StoreAssessmentResultRequest { AssessmentResult result = 1; }
-message StoreAssessmentResultResponse {}
+message StoreAssessmentResultResponse {
+  bool status = 1;
+  string statusMessage = 2;
```
Comment (oxisto) [inline api/orchestrator/orchestrator.proto:244]: `status_message`

File: logging/log.go (hunk 25 above)
Comment (oxisto) [inline logging/log.go:1]: a nice idea, but can we do that in a separate PR? Or is it needed for the tests?

**Review by oxisto — COMMENTED** [review]
Some last comments about variable naming

File: logging/log.go [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+// This file is part of Clouditor Community Edition.
+
+package logging
+
+import (
+	"github.com/sirupsen/logrus"
+	"unicode"
+)
+
+type CapitalizeFormatter struct {
```
Comment (lebogg) [inline logging/log.go:33]:
```suggestion
// CapitalizeFormatter implements the Formatter interface for customizing the logging to our needs
type CapitalizeFormatter struct {
```

File: logging/log.go [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+	logrus.Formatter
+}
+
+func (f CapitalizeFormatter) Format(entry *logrus.Entry) ([]byte, error) {
```
Comment (lebogg) [inline logging/log.go:37]:
```suggestion
// Format capitalizes the first letter of entry's message
func (f CapitalizeFormatter) Format(entry *logrus.Entry) ([]byte, error) {
```

File: logging/log.go [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+	msg := entry.Message
```
Comment (lebogg) [inline logging/log.go:38]: We could also check for special words like gRPC that won't get capitalized. But I think it should not be necessary because we always can/should avoid using such words at the beginning of a logging message.

File: logging/log.go [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+	msgAsRune := []rune(msg)
```
Comment (lebogg) [inline logging/log.go:39]: Instead of message we could use entry.Message directly here?

Comment (lebogg) [inline logging/log.go:39]:
```suggestion
	// Convert message to runes, capitalize first rune and convert back to string again
	msgAsRune := []rune(msg)
```

**Review by lebogg — COMMENTED** [review]
Just looked at the new `log.go` file. Will do a full review later today.

File: logging/log.go (hunk 25 above)
Comment (garuppel) [reply logging/log.go:1]: New files should also have 2016-2020? Not the year they were created?

Comment (lebogg) [reply logging/log.go:1]: yes (2016-2022), I will do a PR to update all headers accordingly.

**Comment by garuppel** [comment]
> Just looked at the new `log.go` file. Will do a full review later today.

I push that in a separate PR.

**Review by oxisto — APPROVED** [review]
